packaging/pep517_backend/_backend.py

```diff
@@ -0,0 +1,108 @@
+"""PEP 517 build backend wrapper for optionally pre-building docs for sdist."""
+
+from __future__ import annotations
+
+import os
+import subprocess
+import sys
+from configparser import ConfigParser
+from importlib.metadata import import_module
+from pathlib import Path
+
+from setuptools.build_meta import (
+    build_sdist as _setuptools_build_sdist,
+    get_requires_for_build_sdist as _setuptools_get_requires_for_build_sdist,
+)
+
+
+__all__ = (  # noqa: WPS317, WPS410
+    'build_sdist', 'get_requires_for_build_sdist',
+)
+
+
+def _make_in_tree_ansible_importable():
+    """Add the library directory to module lookup paths."""
+    lib_path = str(Path.cwd() / 'lib/')
+    os.environ['PYTHONPATH'] = lib_path  # NOTE: for subprocesses
+    sys.path.insert(0, lib_path)  # NOTE: for the current runtime session
+
+
+def _get_package_distribution_version():
+    """Retrieve the current version number from setuptools config."""
+    setup_cfg_path = Path.cwd() / 'setup.cfg'
+    setup_cfg = ConfigParser()
+    setup_cfg.read_string(setup_cfg_path.read_text())
+    cfg_version = setup_cfg.get('metadata', 'version')
+    importable_version_str = cfg_version[len('attr: '):]
+    version_mod_str, version_var_str = importable_version_str.rsplit('.', 1)
+    return getattr(import_module(version_mod_str), version_var_str)
+
+
+def _generate_rst_in_templates():
+    """Create ``*.1.rst.in`` files out of CLI Python modules."""
+    generate_man_cmd = (
+        sys.executable,
+        'hacking/build-ansible.py',
+        'generate-man',
+        '--template-file=docs/templates/man.j2',
+        '--output-dir=docs/man/man1/',
+        '--output-format=man',
+        *Path('lib/ansible/cli/').glob('*.py'),
+    )
+    subprocess.check_call(tuple(map(str, generate_man_cmd)))
+    return Path('docs/man/man1/').glob('*.1.rst.in')
+
+
+def _convert_rst_in_template_to_manpage(rst_in, version_number):
+    """Render pre-made ``*.1.rst.in`` templates into manpages.
+
+    This includes pasting the hardcoded version into the resulting files.
+    The resulting ``in``-files are wiped in the process.
+    """
+    templated_rst_doc = rst_in.with_suffix('')
+    templated_rst_doc.write_text(
+        rst_in.read_text().replace('%VERSION%', version_number))
+
+    rst_in.unlink()
+
+    rst2man_cmd = (
+        sys.executable,
```

Comment:
Why not just do this inline in our interpreter with `docutils.core.publish_file` (or one of the other ones that takes string input, since we've already read the source in to hack on it anyway)? Locating the wrapper script that basically does the same thing feels brittle...

Reply:
Because I was converting a series of calls across makefiles and wasn't implementing this from scratch. So I never checked what APIs are available. FWIW, I was trying to scope this effort to reproduce what makefiles do w/o changes, so it's easier to verify that it's doing the same thing. Optimization/refactoring could be a separate PR IMO.

Reply:
This appears to do the same thing sans subprocess and script:
```
    from docutils.core import publish_file
    from docutils.writers import manpage
    from io import StringIO

    templated_rst_doc = rst_in.read_text().replace('%VERSION%', version_number)
    rst_in.unlink()

    publish_file(source=StringIO(templated_rst_doc), destination_path=rst_in.with_suffix('').with_suffix(''), writer=manpage.Writer())
```

Reply:
@nitzmahone still, I wouldn't want to couple refactoring of the manpage builds with the packaging aspects. How do you feel about having this in a separate PR?

P.S. Pro tip: `StringIO`/`BytesIO` should be used as context managers, otherwise, they need to be closed explicitly to clean up the resources correctly. This is not always a problem, but is a nice principle to follow for consistency with other I/O interactions.

Reply:
I may be mostly projecting my own behavior here :wink:, but "I'll fix it later" will probably be written on my tombstone.

Reply:
For me, it may end up being a never-ending PR refactoring that needs more and more testing, though. It'll end up blocked for longer this way.

Reply:
Oh, and it's not something that this PR changes — if you try pinning setuptools to v39.2.0 on devel, build will traceback too. This is not something that we smoke-test but with this PR, we will. The reason nobody reports this is that PEP 517 pull in the latest setuptools for the given platform.
I can imagine that on some downstreams it might work, maybe, when they call legacy `setup.py` directly or use older Python, so I'm not looking into bumping it in `pyproject.toml`, plus it's out of the scope here anyway. Feature-wise, the currently set value is enough, though.